crates/bevy_ecs/src/schedule_v3/executor/single_threaded.rs

```diff
@@ -0,0 +1,161 @@
+
+use crate::{
+    schedule_v3::{
+        is_apply_system_buffers, ticks_since, ExecutorKind, SystemExecutor, SystemSchedule,
+    },
+    world::World,
+};
+
+/// Runs the schedule using a single thread.
+#[derive(Default)]
+pub struct SingleThreadedExecutor {
+    /// System sets whose conditions have either been evaluated or skipped.
+    completed_sets: FixedBitSet,
+    /// Systems that have run or been skipped.
+    completed_systems: FixedBitSet,
+    /// Systems that have run but have not had their buffers applied.
+    unapplied_systems: FixedBitSet,
+}
+
+impl SystemExecutor for SingleThreadedExecutor {
+    fn kind(&self) -> ExecutorKind {
+        ExecutorKind::SingleThreaded
+    }
+
+    fn init(&mut self, schedule: &SystemSchedule) {
+        // pre-allocate space
+        let sys_count = schedule.system_ids.len();
+        let set_count = schedule.set_ids.len();
+        self.completed_sets = FixedBitSet::with_capacity(set_count);
+        self.completed_systems = FixedBitSet::with_capacity(sys_count);
+        self.unapplied_systems = FixedBitSet::with_capacity(sys_count);
+    }
+
+    fn run(&mut self, schedule: &mut SystemSchedule, world: &mut World) {
+        for sys_idx in 0..schedule.systems.len() {
+            if self.completed_systems.contains(sys_idx) {
+                continue;
+            }
+
+            #[cfg(feature = "trace")]
+            let name = schedule.systems[sys_idx].get_mut().name();
+            #[cfg(feature = "trace")]
+            let should_run_span = info_span!("check_conditions", name = &*name).entered();
+
+            // evaluate conditions
+            let mut should_run = true;
+            let world_tick = world.change_tick();
+
+            // evaluate system set conditions in hierarchical order
+            for set_idx in schedule.sets_of_systems[sys_idx].ones() {
+                if self.completed_sets.contains(set_idx) {
+                    continue;
+                }
+
+                let set_conditions = schedule.set_conditions[set_idx].get_mut();
+
+                // if any condition fails, we need to restore their change ticks
+                let saved_tick = set_conditions
+                    .iter()
+                    .map(|condition| condition.get_last_change_tick())
+                    .min_by_key(|&tick| ticks_since(tick, world_tick));
+
+                let set_conditions_met = set_conditions.iter_mut().all(|condition| {
+                    condition.set_last_change_tick(saved_tick.unwrap());
+                    #[cfg(feature = "trace")]
+                    let _condition_span =
+                        info_span!("condition", name = &*condition.name()).entered();
+                    condition.run((), world)
+                });
+
+                self.completed_sets.insert(set_idx);
+
+                if !set_conditions_met {
+                    // restore change ticks
+                    for condition in set_conditions.iter_mut() {
```

Comment:
I ran into similar concerns when I was writing tests for this. https://github.com/maniwani/bevy/pull/65. There is some inconsistency when resetting change detection with the interaction between system set run criteria and system run criteria.

Reply:
If change ticks aren't reset, how do you write a condition looking for a change since the last time the system (or system set) successfully ran?

Reply:
> I'm guessing the rationale behind this behavior was to propagate changes forward in cases like if `x.is_changed()` && `y.is_changed()`, where only one of the two is changed. 

If the change ticks weren't reset, weird stuff happens when multiple conditions are attached. Let's say we do as you both suggest and tick every condition every frame. Consider this test.

```rust
#[test]
fn pray_the_stars_align() {
    #[derive(Resource)]
    struct X;

    #[derive(Resource, PartialEq)]
    enum Color {
        Red,
        Blue,
    }

    fn changed_system() {
        println!("system ran because X had changed and Color was Red!")
    }

    let mut world = World::new();
    world.insert_resource(X);
    world.insert_resource(Color::Blue);

    let mut schedule = Schedule::new();
    schedule.add_system(changed_system
        .run_if(|x: Res<X>| x.is_changed())
        .run_if(|x: Res<Color>| *color == Color::Red)
    ));

    // X was changed, but color is blue => system does not run
    schedule.run(&mut world);
    
    // change color 
    world.insert_resource(Color::Red);
    
    // color is red, but X no longer seen as changed *by the condition* => system does not run
    schedule.run(&mut world);

    // nothing printed :(
}
```

Does that feel intuitive or consistent?

> Even <resetting> behavior feels wrong. This doesn't behave like the equivalent line in a normal system.

*Technically*, run criteria have never been equivalent to a statement in the system. Half the effect of having one is that a system isn't ticked unless its run criteria passes. It's just that, until now, systems had at most one run criteria, so we never had to address this scenario.

Honestly, before we added `set_last_change_tick`, I was just going (to try) to ban conditions from using change detection entirely. But once we added it, I felt that resetting was at least consistent in the sense of "since last run", but now I see that it breaks when you ask if something *didn't* change.

I don't know what the right answer is.

Reply:
> Does that feel intuitive or consistent? (*I* certainly don't think so.) 
> [...] I don't know what behavior would be the most consistent/intuitive, so if y'all feel strongly about it, I'll change it.

This actually does feel intuitive and consistent (and I feel pretty strongly about it). Run criteria are evaluated every update, just like systems. Just like systems, change detection is tied to each run criteria evaluation. On the first update, the run criteria is evaluated and only one of the criteria is met. On the second update only the other criteria is met. Therefore the system never runs. Easy to "mentally compute" and aligned with user expectations established by systems.

Any other behavior would be a departure from normal per-update change detection behavior. And the mental model introduced by "change detection rollback" is both way harder to reason about for a given context (when it is working as expected) and fully broken (when checking if something hasn't changed).

Reply:
Maybe we could introduce new logic (ex: via a new system param) that accomplishes the "wait until all things have changed"  goal? Ex: first store the current change tick in the system param, then check each resource defined in the system param to see if it has changed. If they have all changed, then return true and set the stored change tick to the current tick. If they have not all changed, return false and leave the tick at the old tick.

Reply:
> Maybe we could introduce new logic that accomplishes the "wait until all things have changed".

I'm not adamant about having this. The main behavior I wanted was "if a system set's conditions fail, we'll skip all systems in it, at once". Hopefully you agree that's better than re-running them for each system in the set, returning different results at different points in the update. 

I think the cumulative/rollback behavior probably stemmed from that.

> This actually does feel intuitive and consistent (and I feel pretty strongly about it). 

Overall, I'm cool with it working like that if you are, but some questions:

- How would you articulate how it works to a user? 
- (If you agreed above), we need to skip the systems in a system set if its conditions return `false`. I think we now have to actively avoid "carrying over results" by making sure to run[^1] every condition every update, even for systems that have already been skipped. Does that make sense? / Seem okay?
- Conditions would be evaluated every update while their guarded system(s) would not be. If a guarded system makes the same change-filtered queries as its conditions, it'll observe all changes that happened in the interim, `>=` those its conditions saw that update.

It's really that last point that has me skeptical. It seems a little like we're taking behavior that was coincidental and promoting it into a rule. 

Imagine you wanted a system's criteria to be "the sky is blue" and "an event is pending". I'd assume that you care about events pending *from the system's POV*, not its conditions'. (Events being double-buffered is an outside detail.)

But again, fine with doing it that way, especially since "did not change" is broken if it's cumulative.

> And the mental model introduced by "change detection rollback" is both way harder to reason about.

They're equally understandable to me, but I'll take your word for it. 
- "Did \<thing happen\> since the last time the guarded system (set) ran?"
- "Did \<thing happen\> since the last time the we checked?" (with conditions evaluated every update)

[^1]: Or just update their change tick.

Reply:
I think it makes sense to always evaluate run criteria. We aren't able to enforce that run criteria don't have local state. So there might be some other local state that needs to be updated every frame. In this model, I think the systems change ticks would need to be updated even if it doesn't run.

Reply:
> The main behavior I wanted was "if a system set's conditions fail, we'll skip all systems in it, at once". Hopefully you agree that's better than re-running them for each system in the set, returning different results at different points in the update. I think the cumulative/rollback behavior probably stemmed from that.

Agreed!

> How would you articulate how it works to a user?

Essentially how you've phrased it:
* Systems run when their conditions evaluate to true
* Conditions are evaluated every frame and their change detection behaves like normal systems running unconditionally every frame.

> (If you agreed above), we need to skip the systems in a system set if its conditions return false. I think we now have to actively avoid "carrying over results" by making sure to run[1](https://github.com/bevyengine/bevy/pull/6587#user-content-fn-1-a1d159e4e4970ca6ae98657794077cfd) every condition every update, even for systems that have already been skipped. Does that make sense? / Seem okay?

That does make sense / seems ok to me. We might be able to get away with short circuiting, as long as we bump the change tick unconditionally for all systems. But probably best to just run everything for now, given that local state + mutations can be a thing. 

> Conditions would be evaluated every update while their guarded system(s) would not be. If a guarded system makes the same change-filtered queries as its conditions, it'll observe all changes that happened in the interim, >= those its conditions saw that update. This is intentional / desirable?

This is a very good question / not something I had considered.

For a given criteria that does something like `if x_changed_query.iter().is_some()` and a system that does some action for all `x_changed_query` items, the data might not line up anymore. A change detected in a run criteria will always show up in a system's change detection, but a system's change detection might contain more matches than a criteria's change detection. For the "did anything change" criteria case (ex: resource changes and `!query.is_empty()`) this won't matter. Because systems will always contain the set of changes matched in their criteria, I think almost all cases will be fine. I can't think of a reasonable corner case example, but theoretically one exists.

But yeah, from my perspective running criteria every frame (or at least, bumping their change tick) is still the least confusing approach, given the outright brokenness of inverted change detection.

> They're equally understandable to me, but I'll take your word for it.
> "Did <thing happen> since the last time the guarded system (set) ran?"

Hmm yeah you're right that from the perspective of a set, the phrasing you used is reasonable + understandable.  I think it gets harder to reason about in the context of a system belonging to multiple sets. As now each set exists at different places in the "change timeline". That feels confusing / hard to compute whether or not a system will run, when compared to all sets existing in the same place on the "change timeline". I know this isn't strictly true as each set criteria will have its own last change tick and sets evaluate at different times in a frame. I'm specifically talking about the problem of a system having multiple criteria, each with ticks from arbitrary frames.

> In this model, I think the systems change ticks would need to be updated even if it doesn't run. Hmm...or maybe not. It could definitely be surprising that if there are changes observed on the system that aren't seen by the run criteria, but the explanation is simple. "The system sees the changes from the last time the system ran.". With updating the change ticks.

Tying all systems' change detection to each update would severely reduce the usability of change detection. Ex: `Changed<T>` queries would be completely unreliable at catching changes in systems that run under a criteria (ex: FixedUpdate criteria). It would drop changes regularly and unpredictably.

Reply:
I think this is resolved now. Executor should run conditions whether or not the system is to be skipped (either we run the conditions and kickoff the system, or we just run the conditions).